File: examples/debugger.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2016 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

#include <pmix_tool.h>

static int attach_to_running_job(char *nspace);

int main(int argc, char **argv)
{
    pmix_status_t rc;
    pmix_proc_t myproc;
    pmix_info_t *info;
    pmix_app_t *app;
    size_t ninfo, napps;
    
    /* check for user directives - this would include:
     * - a flag indicating we want to attach to a specified application
     * - application info if we are launching a new app
     */
    
    /* init us - if a PMIx server pid was provided, then pass it along */
    if (0 < server_pid) {
        ninfo = 1;
        PMIX_INFO_CREATE(info, ninfo);
        PMIX_INFO_LOAD(&info[0], PMIX_SERVER_PIDINFO, server_pid, PMIX_UINT32);
    } else {
        info = NULL;
        ninfo = 0;
    }
    if (PMIX_SUCCESS != (rc = PMIx_tool_init(&myproc, info, ninfo))) {
        fprintf(stderr, "PMIx_tool_init failed: %d\n", rc);
        exit(rc);
    }
    if (0 < ninfo) {
        PMIX_INFO_FREE(info, ninfo);
    }
    fprintf(stderr, "Tool ns %s rank %d: Running\n", myproc.nspace, myproc.rank);

    /* if we are attaching to a running job, then attach to it */
    if (attach) {
        ret = attach_to_running_job(argv[1]);
    } else {
        /* this is an initial launch - we need to launch the application
         * plus the debugger daemons, letting the RM know we are debugging
         * so that it will "pause" the app procs until we are ready */
        napps = 2;
        PMIX_APP_CREATE(app, napps);
        /* setup the executable */
        app[0].cmd = strdup("app");
        app[0].argc = 1;
        app[0].argv = (char**)malloc(2*sizeof(char*));
        app[0].argv[0] = strdup("app");
        app[0].argv[1] = NULL;
        /* provide directives so the apps do what the user requested */
        ninfo = 2;
        PMIX_INFO_CREATE(app[0].info, ninfo);
        PMIX_INFO_LOAD(&app[0].info[0], PMIX_NP, 128, PMIX_UINT64);
        PMIX_INFO_LOAD(&app[0].info[0], PMIX_MAPBY, "slot", PMIX_STRING);

        /* setup the name of the daemon executable to launch */
        app[1].cmd = strdup("debuggerdaemon");
        app[1].argc = 1;
        app[1].argv = (char**)malloc(2*sizeof(char*));
        app[1].argv[0] = strdup("debuggerdaemon");
        app[1].argv[1] = NULL;
        /* provide directives so the daemons go where we want, and
         * let the RM know these are debugger daemons */
        ninfo = 2;
        PMIX_INFO_CREATE(app[1].info, ninfo);
        PMIX_INFO_LOAD(&app[1].info[0], PMIX_MAPBY, "ppr:1:node", PMIX_STRING);  // instruct the RM to launch one copy of the executable on each node
        PMIX_INFO_LOAD(&app[1].info[1], PMIX_DEBUGGER_DAEMONS, true, PMIX_BOOL); // these are debugger daemons
        /* spawn the daemons */
        PMIx_Spawn(NULL, 0, app, napps, dspace);
        /* cleanup */
        PMIX_APP_FREE(app, napps);
        
        /* this is where a debugger tool would wait until the debug operation is complete */
    }

    
 done:
    PMIx_tool_finalize(NULL, 0);

    return(ret);
}

static int attach_to_running_job(char *nspace)
{
    pmix_status_t rc;
    pmix_proc_t myproc;
    pmix_info_t *info;
    pmix_app_t *app;
    size_t ninfo, napps;

    /* query the active nspaces so we can verify that the
     * specified one exists */
    ninfo = 1;
    PMIX_INFO_CREATE(info, ninfo);
    (void)strncpy(info[0].key, PMIX_QUERY_NAMESPACES, PMIX_MAX_KEYLEN);
    if (PMIX_SUCCESS != (rc = PMIx_Query_info(info, ninfo))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Query_info failed: %d\n", myproc.nspace, myproc.rank, rc);
        return -1;
    }
    /* the query should have returned a comma-delimited list of nspaces */
    if (PMIX_STRING != info[0].type) {
        fprintf(stderr, "Query returned incorrect data type: %d\n", info[0].type);
        return -1;
    }
    if (NULL == info[0].data.string) {
        fprintf(stderr, "Query returned no active nspaces\n");
        return -1;
    }
    /* split the returned string and look for the given nspace */

    /* if not found, then we have an error */
    PMIX_INFO_FREE(info, ninfo);
    
    /* get the proctable for this nspace */
    ninfo = 1;
    PMIX_INFO_CREATE(info, ninfo);
    (void)strncpy(info[0].key, PMIX_QUERY_PROC_TABLE, PMIX_MAX_KEYLEN);
    (void)strncpy(info[0].qualifier, nspace, PMIX_MAX_KEYLEN);
    if (PMIX_SUCCESS != (rc = PMIx_Query_info(info, ninfo))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Query_info failed: %d\n", myproc.nspace, myproc.rank, rc);
        return -1;
    }
    /* the query should have returned a data_array */
    if (PMIX_DATA_ARRAY != info[0].type) {
        fprintf(stderr, "Query returned incorrect data type: %d\n", info[0].type);
        return -1;
    }
    if (NULL == info[0].data.darray.array) {
        fprintf(stderr, "Query returned no proctable info\n");
        return -1;
    }
    /* the data array consists of a struct:
     *     size_t size;
     *     void* array;
     *
     * In this case, the array is composed of pmix_proc_info_t structs:
     *     pmix_proc_t proc;   // contains the nspace,rank of this proc
     *     char* hostname;
     *     char* executable_name;
     *     pid_t pid;
     *     int exit_code;
     *     pmix_proc_state_t state;
     */
    
    /* this is where a debugger tool would process the proctable to
     * create whatever blob it needs to provide to its daemons */
    PMIX_INFO_FREE(info, ninfo);

    /* setup the debugger daemon spawn request */
    napps = 1;
    PMIX_APP_CREATE(app, napps);
    /* setup the name of the daemon executable to launch */
    app[0].cmd = strdup("debuggerdaemon");
    app[0].argc = 1;
    app[0].argv = (char**)malloc(2*sizeof(char*));
    app[0].argv[0] = strdup("debuggerdaemon");
    app[0].argv[1] = NULL;
    /* provide directives so the daemons go where we want, and
     * let the RM know these are debugger daemons */
    ninfo = 3;
    PMIX_INFO_CREATE(app[0].info, ninfo);
    PMIX_INFO_LOAD(&app[0].info[0], PMIX_MAPBY, "ppr:1:node", PMIX_STRING);  // instruct the RM to launch one copy of the executable on each node
    PMIX_INFO_LOAD(&app[0].info[1], PMIX_DEBUGGER_DAEMONS, true, PMIX_BOOL); // these are debugger daemons
    PMIX_INFO_LOAD(&app[0].info[2], PMIX_DEBUG_TARGET, nspace, PMIX_STRING); // the "jobid" of the application to be debugged

    /* spawn the daemons */
    PMIx_Spawn(NULL, 0, app, napps, dspace);
    /* cleanup */
    PMIX_APP_FREE(app, napps);

    /* this is where a debugger tool would wait until the debug operation is complete */

    return 0;
}

File: examples/python/debugger.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/debugger.c
#
# The debugger "front end". On an initial launch it asks the RM what it
# supports (co-spawn? stop-on-exec?), launches the application paused for
# attach, then launches one debugger daemon per node pointed at the
# application's namespace. With --attach <nspace> it instead verifies
# that the named job exists.

import os
import sys

from examples import *

tool = PMIxTool()
myproc = {}
waiting_for_debugger = MyLock()


# this is a callback function for the query API. The query will call back
# with a status indicating whether the request could be fully satisfied,
# partially satisfied, or completely failed; each returned entry carries
# the key that was provided in the query.
def querycbfunc(status, results, mq):
    for item in results:
        eprint("Transferring %s" % as_key(item['key']))
    mq.info = results
    # release the block
    mq.wakeup(status)


# this is the event notification function we pass down below when
# registering for general events - i.e., the default handler
def notification_fn(evhdlr, status, source, info, results):
    return PMIX_EVENT_ACTION_COMPLETE, None


# registered for the debugger job's termination
def release_fn(evhdlr, status, source, info, results):
    waiting_for_debugger.wakeup()
    return PMIX_EVENT_ACTION_COMPLETE, None


def attach_to_running_job(nspace):
    """Verify that the named namespace exists."""
    q = MyQuery()
    rc = tool.query_nb([{'keys': [PMIX_QUERY_NAMESPACES],
                         'qualifiers': None}], querycbfunc, q)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d: PMIx_Query_info failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
        return -1
    q.wait()

    if not q.info:
        eprint("Query returned no info")
        return -1
    # the query should have returned a comma-delimited list of nspaces
    if PMIX_STRING != q.info[0]['val_type']:
        eprint("Query returned incorrect data type: %d"
               % q.info[0]['val_type'])
        return -1
    if q.info[0]['value'] is None:
        eprint("Query returned no active nspaces")
        return -1

    eprint("Query returned %s" % q.info[0]['value'])
    return 0


def spawn_debugger(appspace):
    """Launch one debugger daemon per node, pointed at appspace."""
    cwd = os.getcwd()
    debuggerd = os.path.join(cwd, "python", "debuggerd.py") \
        if os.path.isdir(os.path.join(cwd, "python")) \
        else os.path.join(os.path.dirname(os.path.abspath(__file__)),
                          "debuggerd.py")
    debugger = [{'cmd': debuggerd, 'argv': [debuggerd], 'cwd': cwd,
                 'maxprocs': 1, 'env': None, 'info': []}]
    # provide directives so the daemons go where we want, and let the RM
    # know these are debugger daemons
    dinfo = [
        # instruct the RM to launch one copy of the executable on each node
        {'key': PMIX_MAPBY, 'value': "ppr:1:node", 'val_type': PMIX_STRING},
        # these are debugger daemons
        {'key': PMIX_DEBUGGER_DAEMONS, 'value': True,
         'val_type': PMIX_BOOL},
        # the nspace being debugged
        {'key': PMIX_DEBUG_JOB, 'value': appspace,
         'val_type': PMIX_STRING},
        # notify us when the debugger job completes
        {'key': PMIX_NOTIFY_COMPLETION, 'value': True,
         'val_type': PMIX_BOOL}]
    # spawn the daemons
    eprint("Debugger: spawning %s" % debugger[0]['cmd'])
    rc, dspace = tool.spawn(dinfo, debugger)
    if PMIX_SUCCESS != rc:
        eprint("Debugger daemons failed to launch with error: %s"
               % tool.error_string(rc))
    eprint("SPAWNED DEBUGGERD")
    return rc


def main():
    global myproc

    nspace = None

    # Process any arguments we were given
    i = 1
    while i < len(sys.argv):
        arg = sys.argv[i]
        if arg in ("-h", "--help"):
            # print the usage message and exit
            pass
        if arg in ("-a", "--attach"):
            if nspace is not None:
                # can only support one
                eprint("Cannot attach to more than one nspace")
                sys.exit(1)
            # the next argument must be the nspace
            i += 1
            if i == len(sys.argv):
                # they goofed
                eprint("The %s option requires an <nspace> argument" % arg)
                sys.exit(1)
            nspace = sys.argv[i]
        else:
            eprint("Unknown option: %s" % arg)
            sys.exit(1)
        i += 1

    # use the system connection first, if available. Note that the C
    # version builds this directive but then passes ninfo == 0, so the
    # library never sees it - the port keeps that behavior
    info = []

    # init as a tool
    rc, myproc = tool.init(info)
    if PMIX_SUCCESS != rc:
        eprint("PMIx_tool_init failed: %d" % rc)
        sys.exit(rc)

    eprint("Tool ns %s rank %d: Running"
           % (myproc['nspace'], myproc['rank']))

    # register a default event handler
    tool.register_event_handler(None, None, notification_fn)

    # register another handler specifically for when the debugger job
    # completes
    tool.register_event_handler([PMIX_ERR_JOB_TERMINATED], None, release_fn)

    # if we are attaching to a running job, then attach to it
    if nspace is not None:
        rc = attach_to_running_job(nspace)
        if PMIX_SUCCESS != rc:
            eprint("Failed to attach to nspace %s: error code %d"
                   % (nspace, rc))
            return done(rc)
        return done(rc)

    # this is an initial launch - we need to launch the application plus
    # the debugger daemons, letting the RM know we are debugging so that
    # it will "pause" the app procs until we are ready. First we need to
    # know if this RM supports co-spawning of daemons with the
    # application, or if we need to launch the daemons as a separate spawn
    # command. The former is faster and more scalable, but not every RM
    # may support it. We also need to ask for debug support so we know if
    # the RM can stop-on-exec, or only supports stop-in-init
    myquery_data = MyQuery()
    eprint("Debugger: querying capabilities")
    rc = tool.query_nb([{'keys': [PMIX_QUERY_SPAWN_SUPPORT,
                                  PMIX_QUERY_DEBUG_SUPPORT],
                         'qualifiers': None}], querycbfunc, myquery_data)
    if PMIX_SUCCESS != rc:
        eprint("PMIx_Query_info failed: %d" % rc)
        return done(rc)
    myquery_data.wait()

    # we should have received back two info structs, one containing a
    # comma-delimited list of PMIx spawn attributes the RM supports, and
    # the other containing a comma-delimited list of PMIx debugger
    # attributes it supports
    if 2 != len(myquery_data.info):
        # this is an error
        eprint("PMIx Query returned an incorrect number of results: %lu"
               % len(myquery_data.info))
        return done(rc)

    # we would like to co-spawn the debugger daemons with the app, but
    # let's first check to see if this RM supports that operation by
    # looking for the PMIX_COSPAWN_APP attribute in the spawn support.
    #
    # We will also check to see if "stop_on_exec" is supported. Few RMs
    # do so, which is why we have to check.
    #
    # Note that the PMIx reference server always returns the query results
    # in the same order as the query keys. However, this is not
    # guaranteed, so we search the returned info for the desired key
    cospawn = False
    stop_on_exec = False
    for item in myquery_data.info:
        if key_is(item['key'], PMIX_QUERY_SPAWN_SUPPORT):
            # see if the cospawn attribute is included
            cospawn = as_key(PMIX_COSPAWN_APP) in item['value']
        elif key_is(item['key'], PMIX_QUERY_DEBUG_SUPPORT):
            stop_on_exec = as_key(PMIX_DEBUG_STOP_ON_EXEC) in item['value']

    # if cospawn is true, then we can launch both the app and the debugger
    # daemons at the same time
    if cospawn:
        pass
    else:
        # we must do these as separate launches, so do the app first
        cwd = os.getcwd()   # point us to our current directory
        app = [{'cmd': "client", 'argv': ["./client"], 'cwd': cwd,
                'maxprocs': 2, 'env': None, 'info': []}]
        # provide job-level directives so the apps do what the user
        # requested
        info = [
            # map by slot
            {'key': PMIX_MAPBY, 'value': "slot", 'val_type': PMIX_STRING},
            # procs are to stop on first instruction, or else pause in
            # init for debugger attach
            {'key': PMIX_DEBUG_STOP_ON_EXEC if stop_on_exec
             else PMIX_DEBUG_STOP_IN_INIT, 'value': True,
             'val_type': PMIX_BOOL},
            # forward stdout and stderr to me
            {'key': PMIX_FWD_STDOUT, 'value': True, 'val_type': PMIX_BOOL},
            {'key': PMIX_FWD_STDERR, 'value': True, 'val_type': PMIX_BOOL}]

        # spawn the job - the call returns when the app has been launched
        eprint("Debugger: spawning %s" % app[0]['cmd'])
        rc, appspace = tool.spawn(info, app)
        if PMIX_SUCCESS != rc:
            eprint("Application failed to launch with error: %s(%d)"
                   % (tool.error_string(rc), rc))
            return done(rc)

        # now launch the debugger daemons
        rc = spawn_debugger(appspace)
        if PMIX_SUCCESS != rc:
            return done(rc)

    # this is where a debugger tool would wait until the debug operation
    # is complete
    waiting_for_debugger.wait()

    return done(rc)


def done(rc):
    tool.finalize()
    return rc


if __name__ == '__main__':
    sys.exit(main())
